File: TUI-ConsoleLauncher/app/src/main/java/ohi/andre/consolelauncher/ai/MiniMaxService.java
package ohi.andre.consolelauncher.ai;

import android.content.Context;
import android.util.Log;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;

/**
 * MiniMax Service - API client for MiniMax AI
 * Handles HTTP communication with MiniMax Chat API
 */
public class MiniMaxService {

    private static final String TAG = "MiniMaxService";
    private static final String BASE_URL = "https://api.minimax.chat";
    private static final String CHAT_ENDPOINT = "/v1/chat/completions";
    private static final int TIMEOUT_SECONDS = 30;

    private final String apiKey;
    private final String groupId;
    private final String model;
    private final OkHttpClient httpClient;

    /**
     * Create service with default config
     */
    public MiniMaxService(Context context) {
        this.apiKey = "";
        this.groupId = "";
        this.model = "abab6.5-chat";
        this.httpClient = createHttpClient();
    }

    /**
     * Create service with configuration
     */
    public MiniMaxService(Context context, String apiKey, String groupId) {
        this.apiKey = apiKey;
        this.groupId = groupId;
        this.model = "abab6.5-chat";
        this.httpClient = createHttpClient();
    }

    private OkHttpClient createHttpClient() {
        return new OkHttpClient.Builder()
                .connectTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
                .readTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
                .writeTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
                .addInterceptor(chain -> {
                    Request original = chain.request();
                    Request.Builder builder = original.newBuilder()
                            .header("Content-Type", "application/json")
                            .header("Authorization", "Bearer " + apiKey)
                            .header("X-GroupId", groupId);
                    return chain.proceed(builder.build());
                })
                .build();
    }

    /**
     * Test connection to API
     */
    public MiniMaxResponse testConnection() {
        MiniMaxResponse response = new MiniMaxResponse();

        try {
            RequestBody body = RequestBody.create(
                    "{\"model\":\"" + model + "\",\"messages\":[{\"role\":\"user\",\"content\":\"test\"}],\"max_tokens\":1}",
                    MediaType.parse("application/json")
            );

            Request request = new Request.Builder()
                    .url(BASE_URL + CHAT_ENDPOINT)
                    .post(body)
                    .build();

            Response httpResponse = httpClient.newCall(request).execute();
            response.success = httpResponse.isSuccessful();
            response.message = httpResponse.isSuccessful() ? "Connection successful" : "Error: " + httpResponse.code();
            httpResponse.close();

        } catch (Exception e) {
            Log.e(TAG, "Connection test failed", e);
            response.success = false;
            response.message = e.getMessage();
        }

        return response;
    }

    /**
     * Send chat request
     */
    public MiniMaxResponse sendChatRequest(String message, String systemPrompt, int maxTokens, float temperature) {
        MiniMaxResponse response = new MiniMaxResponse();

        try {
            StringBuilder json = new StringBuilder();
            json.append("{\"model\":\"").append(model).append("\",");
            json.append("\"messages\":[");

            if (systemPrompt != null && !systemPrompt.isEmpty()) {
                json.append("{\"role\":\"system\",\"content\":\"").append(escapeJson(systemPrompt)).append("\"},");
            }

            json.append("{\"role\":\"user\",\"content\":\"").append(escapeJson(message)).append("\"}]");
            json.append(",\"max_tokens\":").append(maxTokens);
            json.append(",\"temperature\":").append(temperature);
            json.append("}");

            RequestBody body = RequestBody.create(json.toString(), MediaType.parse("application/json"));

            Request request = new Request.Builder()
                    .url(BASE_URL + CHAT_ENDPOINT)
                    .post(body)
                    .build();

            Response httpResponse = httpClient.newCall(request).execute();

            if (httpResponse.isSuccessful() && httpResponse.body() != null) {
                String responseBody = httpResponse.body().string();
                response.success = true;
                response.message = extractContent(responseBody);
            } else {
                response.success = false;
                response.message = "API Error: " + httpResponse.code();
            }

            httpResponse.close();

        } catch (Exception e) {
            Log.e(TAG, "Chat request failed", e);
            response.success = false;
            response.message = e.getMessage();
        }

        return response;
    }

    /**
     * Get available models
     */
    public String[] getAvailableModels() {
        return new String[]{
                "abab6.5-chat",
                "abab6.5s-chat",
                "abab5.5-chat"
        };
    }

    /**
     * Extract content from API response
     */
    private String extractContent(String responseBody) {
        try {
            int contentStart = responseBody.indexOf("\"content\":\"");
            if (contentStart == -1) return responseBody;

            contentStart += 11;
            int contentEnd = responseBody.indexOf("\"", contentStart);
            if (contentEnd == -1) return responseBody;

            String content = responseBody.substring(contentStart, contentEnd);
            return content.replace("\\n", "\n").replace("\\\"", "\"");
        } catch (Exception e) {
            return responseBody;
        }
    }

    /**
     * Escape special characters for JSON
     */
    private String escapeJson(String text) {
        if (text == null) return "";
        return text.replace("\\", "\\\\")
                   .replace("\"", "\\\"")
                   .replace("\n", "\\n")
                   .replace("\r", "\\r")
                   .replace("\t", "\\t");
    }

    /**
     * Response container
     */
    public static class MiniMaxResponse {
        public boolean success;
        public String message;
    }
}

File: tui-news-module/tui/feature/news/MiniMaxService.kt
package tui.feature.news

import android.content.Context
import android.util.Log
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import tui.feature.news.models.ChatMessage
import tui.feature.news.models.MiniMaxRequest
import tui.feature.news.models.MiniMaxResponse
import com.google.gson.Gson
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Service layer for MiniMax API communication
 * Handles all network requests to the MiniMax AI platform
 */
class MiniMaxService(private val context: Context) {

    companion object {
        private const val TAG = "MiniMaxService"
        private const val BASE_URL = "https://api.minimax.chat/v1"
        private const val TIMEOUT_SECONDS = 30L
    }

    private val gson = Gson()
    private val client: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
            .readTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
            .writeTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
            .build()
    }

    private val apiKey: String by lazy {
        NewsConfig.getApiKey(context)
    }

    /**
     * Fetches news summary from MiniMax AI
     * @param topic The news topic to query
     * @param detailed Whether to provide detailed summaries
     * @return Result containing the news text or an error
     */
    suspend fun getNewsSummary(topic: String, detailed: Boolean = false): Result<String> {
        return withContext(Dispatchers.IO) {
            try {
                val systemPrompt = buildSystemPrompt(detailed)
                val userPrompt = buildUserPrompt(topic, detailed)

                val messages = listOf(
                    ChatMessage(role = "system", content = systemPrompt),
                    ChatMessage(role = "user", content = userPrompt)
                )

                val request = MiniMaxRequest(messages = messages)
                val jsonBody = gson.toJson(request)

                val mediaType = "application/json".toMediaType()
                val requestBody = jsonBody.toRequestBody(mediaType)

                val httpRequest = Request.Builder()
                    .url("$BASE_URL/text/chatcompletion_v2")
                    .addHeader("Authorization", "Bearer $apiKey")
                    .addHeader("Content-Type", "application/json")
                    .post(requestBody)
                    .build()

                val response = client.newCall(httpRequest).execute()

                if (response.isSuccessful) {
                    val responseBody = response.body?.string()
                    if (responseBody != null) {
                        val miniMaxResponse = gson.fromJson(responseBody, MiniMaxResponse::class.java)
                        val content = miniMaxResponse.choices?.firstOrNull()?.message?.content
                        if (content != null) {
                            Result.success(content)
                        } else {
                            Result.failure(IOException("Empty response from MiniMax API"))
                        }
                    } else {
                        Result.failure(IOException("Empty response body"))
                    }
                } else {
                    val errorMsg = "API Error: ${response.code} - ${response.message}"
                    Log.e(TAG, errorMsg)
                    Result.failure(IOException(errorMsg))
                }
            } catch (e: Exception) {
                Log.e(TAG, "Network error: ${e.message}", e)
                Result.failure(e)
            }
        }
    }

    /**
     * Builds the system prompt that defines the AI's behavior as a news aggregator
     */
    private fun buildSystemPrompt(detailed: Boolean): String {
        return if (detailed) {
            """
            You are a professional news analyst for a terminal-based interface. 
            Your output must be plain text optimized for CLI display.
            
            Rules:
            - NO markdown formatting (no **, ##, >, etc.)
            - Use simple ASCII characters for formatting
            - Maximum 60 characters per line
            - Provide 3-5 in-depth news stories with context
            - Each story should have: headline, key details, and significance
            - Separate stories with blank lines
            - End with source attribution if available
            """.trimIndent()
        } else {
            """
            You are a concise news aggregator for a terminal-based interface.
            Your output must be plain text optimized for CLI display.
            
            Rules:
            - NO markdown formatting (no **, ##, >, etc.)
            - Use simple bullet points (•) or hyphens (-)
            - Maximum 50 characters per line
            - Provide 5-8 brief news headlines with one-line summaries
            - Keep each summary under 20 words
            - Group related stories under topic headers
            """.trimIndent()
        }
    }

    /**
     * Builds the user prompt based on the requested news topic
     */
    private fun buildUserPrompt(topic: String, detailed: Boolean): String {
        val depth = if (detailed) "Provide detailed analysis" else "Brief summaries"
        return "Get the latest news about '$topic'. $depth. Focus on recent developments from the last 24-48 hours."
    }

    /**
     * Validates the API key configuration
     */
    fun validateApiKey(): Boolean {
        return apiKey.isNotBlank() && apiKey.length > 10
    }
}
